### PR TITLE
Server chosen tcpfwd ports

### PR DESCRIPTION
**Pull request by houseofkodai**
Server chosen tcpfwd ports

### CODE REVIEW

File: svr-tcpfwd.c [hunk 1]
```diff
@@ -210,8 +216,16 @@ static int svr_remotetcpreq() {
 		 * has to remember it if it's to be cancelled */
 		m_free(request_addr);
 		m_free(tcpinfo);
+	} else {
```
Comment (mkj) [inline svr-tcpfwd.c]: Can you move this bit this above "out:" ?

Don't think logging should be turned on generally. If you have a particular use case maybe make it a options.h option.

File: netio.c [hunk 2]
```diff
@@ -451,6 +505,8 @@ int dropbear_listen(const char* address, const char* port,
 			continue;
 		}
 
+		if (0 == allocated_lport) allocated_lport = get_sock_port(sock, 1);
```
Comment (mkj) [inline netio.c]: All "if" statements should have {} braces

File: netio.c [hunk 3]
```diff
@@ -400,11 +438,27 @@ int dropbear_listen(const char* address, const char* port,
...
+	 */
+	u_int16_t *allocated_lport_p = 0;
+	int allocated_lport = 0;
 
 	nsock = 0;
 	for (res = res0; res != NULL && nsock < sockcount;
 			res = res->ai_next) {
 
+		if (AF_INET == res->ai_family) {
+			allocated_lport_p = &((struct sockaddr_in *)res->ai_addr)->sin_port;
```
Comment (mkj) [inline netio.c]: allocated_lport_p seems too tricky. Just go like this?
```
if allocated_port 
    if AF_INET 
        .....->sin_port = htons(...)
    else if AF_INET6 {
      .... -> sin6_port = htons(...)

```
?

Also note typo AF_INET not AF_INET6

File: netio.c [hunk 4]
```diff
@@ -348,6 +348,44 @@ void set_sock_priority(int sock, enum dropbear_prio prio) {
 
 }
 
+/* from openssh/canohost.c avoid premature-optimization */
+int get_sock_port(int sock, int local) {
```
Comment (mkj) [inline netio.c]: Nothing calls this with local=0, can you remove that case?

**Comment by houseofkodai** [comment]
Dear mkj,

apologies for the delay - all changes updated in the new commits.

- ka

File: netio.c (hunk 4 above)
Comment (houseofkodai) [inline netio.c]: my bad, done.

File: netio.c (hunk 3 above)
Comment (houseofkodai) [inline netio.c]: simplified logic and fixed typo.

File: netio.c (hunk 2 above)
Comment (houseofkodai) [inline netio.c]: done.

File: svr-tcpfwd.c (hunk 1 above)
Comment (houseofkodai) [inline svr-tcpfwd.c]: done.

**Pull request merged**